tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc

```diff
@@ -0,0 +1,242 @@
+      b->BuildAttrValue(func_, &f_attr);
+
+      AttrValue other_arguments_types_attr;
+      b->BuildAttrValue(other_arguments_types, &other_arguments_types_attr);
+
+      AttrValue preserve_cardinality_attr;
+      b->BuildAttrValue(preserve_cardinality_, &preserve_cardinality_attr);
+
+      TF_RETURN_IF_ERROR(b->AddDataset(
+            this, {std::make_pair(0, input_node)},
+            {std::make_pair(1, other_arguments)},
+            {std::make_pair("predicate", f_attr),
+             std::make_pair("Targuments", other_arguments_types_attr),
+             std::make_pair("preserve_cardinality", 
+                            preserve_cardinality_attr)},
+             output));
+      return Status::OK();
+    }
+
+   private:
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {}
+
+      Status Initialize(IteratorContext* ctx) override {
+        TF_RETURN_IF_ERROR(
+            dataset()->input_->MakeIterator(ctx, prefix(), &input_impl_));
+        return dataset()->captured_func_->Instantiate(
+            ctx, &instantiated_captured_func_);
+      }
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        {
+          tf_shared_lock l(mu_);
+          if(!input_impl_) {
+            *end_of_sequence = true;
+            return Status::OK();
+          }
+          TF_RETURN_IF_ERROR(
+              input_impl_->GetNext(ctx, out_tensors, end_of_sequence));
+        }
+        if (*end_of_sequence) {
+          mutex_lock l(mu_);
+          input_impl_.reset();
+          return Status::OK();
+        }
+
+        std::vector<Tensor> bool_output;
+
+        Status s = instantiated_captured_func_->RunWithBorrowedArgs(
+            ctx, *out_tensors, &bool_output);
+
+        if (s.ok()) {
+          if(bool_output.size() != 1 || bool_output[0].dtype() != DT_BOOL ||
+             bool_output[0].NumElements() != 1) {
+            return errors::InvalidArgument(
+                "`predicate` must returns a scalar bool tensor.");
+          }
+          auto cond = bool_output[0].scalar<bool>()();
+          if (!cond) { // predicate is false
+            *end_of_sequence = true;
+            return Status::OK();
+          }
+        } else if (errors::IsOutOfRange(s)) {
+          if (dataset()->preserve_cardinality_) {
+            // To guarantee that the transformation preserves the cardinality of
+            // the dataset, we convert `OutOfRange` to `InvalidArgument` as the
+            // former may be interpreted by a caller as the end of sequence.
+            return errors::InvalidArgument(
+                "Function invocation produced OutOfRangeError: ",
+                s.error_message());
+          } else {
+            // `f` may deliberately raise `errors::OutOfRange` to indicate
```

Comment:
I think this might be true for the function passed to `map` but the contract for the function passed to `take_while` should be that early termination is indicated by returning `false`. So if the `take_while` predicate returns an error, I think it makes sense to simply propagate the error to the caller.